tmk_core/common/avr/bootloader.c

```diff
@@ -245,6 +245,44 @@ __attribute__((weak)) void bootloader_jump(void) {
 #    endif
                    [bootaddrme] "M"((((FLASH_SIZE - BOOTLOADER_SIZE) >> 1) >> 8) & 0xff), [bootaddrlo] "M"((((FLASH_SIZE - BOOTLOADER_SIZE) >> 1) >> 0) & 0xff));
 
+#elif defined(BOOTLOADER_NANOBOOT)
+    /* nanoBoot sets up 2 registers (r2,r3), copies then clears the MCUSR register,
+    disables the watchdog timer and then checks the EXTRF. It'll only continue in the
+    bootloader if the flag is set, otherwise it'll boot the application (jmp 0) when
+    using the reset vector (0x7E00). This code jumps past the reset vector straight into
+    the bootloader skipping the EXTRF check, so prior setup is required. */
+    cli();
+    UDCON  = 1;
+    USBCON = (1 << FRZCLK);  // disable USB
+    UCSR1B = 0;
+    _delay_ms(1000);  // delay seemed to help but issue not limited to hid_listen
+    EIMSK  = 0;
+    PCICR  = 0;
+    SPCR   = 0;
+    ACSR   = 0;
+    EECR   = 0;
+    ADCSRA = 0;
+    TIMSK0 = 0;
+    TIMSK1 = 0;
+    TIMSK3 = 0;
+    TIMSK4 = 0;
+    UCSR1B = 0;
+    TWCR   = 0;
+    DDRB   = 0;
+    DDRC   = 0;
+    DDRD   = 0;
+    DDRE   = 0;
+    DDRF   = 0;
+    TWCR   = 0;
+    PORTB  = 0;
+    PORTC  = 0;
+    PORTD  = 0;
+    PORTE  = 0;
+    PORTF  = 0;
+    asm volatile("clr r2");      // clear register (rZERO in bootloader)
+    asm volatile("clr r3");      // clear register (rONE in bootloader)
+    asm volatile("inc r3");      // increment register
+    asm volatile("jmp 0x7e6a");  // jump to bootloader skipping some setup and extrf check.
```

Comment:
This address won't work with atmega16u4, it is also wrong for sigprof's nanoboot edit (0x7e68). So this is going to need to definable I guess?

Reply:
Something like this? 
```suggestion
#if defined(__AVR_ATmega16U4__)
    asm volatile("jmp 0x7e6a");  // jump to bootloader skipping some setup and extrf check.
#elif defined(__AVR_ATmega32U4__)
    asm volatile("jmp 0x7e6a");  // jump to bootloader skipping some setup and extrf check.
#else
    // some magic code, using a define maybe?
#endif
```

Reply:
I did not consider this aspect when modifying nanoBoot. Basically, that bootloader does not have proper support for external entry, unless some hack like this one is used, and such hacks are often fragile. E.g., apparently the HID report descriptor used by the LUFA HID bootloader may have a bug (the `Logical Maximum (255)` item represented as 1 byte might be questionable, because some code may treat that value as signed), but fixing that bug may change the descriptor length, therefore moving some nanoBoot code around.

So the possible options here are:

1) Try to support nanoBoot as it is, avoid modifications that change the jump address. (E.g., I could just avoid the optimization that saves 2 bytes by changing the code in that area.)

2) Use another jump address — the earliest possible one would be just after reading the MCUSR register:
    ```
      ; Create a copy of the MCUSR (MCU Status Register) as soon as possible
      in          rMCUSR, _SFR_IO_ADDR(MCUSR)   ; Load MCU Status Register to rMCUSR
    7e5a:       44 b6           in      r4, 0x34        ; 52
      out         _SFR_IO_ADDR(MCUSR), rZERO    ; Load MCU Status Register with rZERO (clear reset flags, particularly clear WDRF in MCUSR), necessary before disabling the Watchdog
    7e5c:       24 be           out     0x34, r2        ; 52
    ```
    Jumping to `0x7e5c` after setting up the appropriate fake value in `rMCUSR` (`r4`) could give the desired result. This address is not yet touched by any optimizations.
    This still won't be 100% safe, because some changes (e.g., the HID report descriptor change mentioned above) may still require moving that code around.

3) Fork nanoBoot and implement support for external bootloader entry (e.g., similar to what is done by ubaboot); support only this fork, and not the original version.

4) Fork ubaboot and implement the same HID protocol there instead of its native protocol, support that bootloader instead of nanoBoot (in that case forced bootloader entry would work even with the original ubaboot, only flashing would require a different ubaboot-specific tool).

Reply:
With the modified hid ubaboot having working eeprom it’s likely the way forward. I’ve already modified hid_bootloader_cli for eeprom support (needs some tidying). 

It also addresses some of the issues that nanoboot has, bootloader overwrite protection and bootloader entry. (Which admittedly I’ve tested a fixed version of nanoboot that addresses this)

I do think the ubaboot fork should have a different name if possible just to stop any issues with people flashing the wrong version. Although that’d be more obvious in this case due to supporting hid with the modified version.